### ISSUE TITLE
MSU-1 implentation needs updated, standalone version skips and pops much less for select games

### ISSUE THREAD

**Issue by prataczak** [libretro/snes9x#70]
I've noticed that some of the games still pop/skip quite a bit, but this is not the case with the most recent build of SNES9x with MSU support (downloadable here:  http://dl.qwertymodo.com/snes9x.zip )

The game in particular that made me decide to test the RetroArch core against the SNES9x standalone version was the recent "Street Fighter 2 Turbo MSU Hack." The youtube page is here, and it has links to all of the required files for the MSU enhancements. https://www.youtube.com/watch?v=Ji3G3GKH5X4

If you start a 1 player game and choose Ryu, when you go to the Zangief stage for your first fight, the track skips noticeably. Also, if you press start to skip the 'capcom' sound effects to the title screen when you first boot the game.

A lot of improvements have been made to the code by qwertymodo apparently. Everything works beautifully on his most recent build.

I think if someone were to update the code for MSU support one more time, it would be perfect and probably wouldn't need tweaked again.

**Comment by hizzlekizzle** [comment]
Does qwertymodo's build perform differently from the regular snes9x 1.54.1 release?

**Comment by prataczak** [comment]
Yes, the regular 1.54.1 doesn't have MSU support.

According to qwertymodo, there have been 17 MSU-1 related commits since January (the last time it was updated), several of which specifically address skipping/popping audio.

**Comment by qwertymodo** [comment]
1.54.1 doesn't support MSU-1 at all, so... yes my build performs differently.  I'm not able to create a formal pull request at the moment, but here are the commits involving MSU-1 audio since the latest upstream sync here (not including the compressed .msu1 pack support, which might also be nice to have).

https://github.com/snes9xgit/snes9x/commit/0df033d3b749d6e121807e1d1a72f85d617c6128
https://github.com/snes9xgit/snes9x/commit/b3b38a6d1adebeed870b093448643e8298588411
https://github.com/snes9xgit/snes9x/commit/32f70fa38e96b90469baa470dac10ffcf186f073
https://github.com/snes9xgit/snes9x/commit/b230e50ec51719f872ce6821d94b003d6dbca383
https://github.com/snes9xgit/snes9x/commit/75037ef380e3a9fbb097a37d6d01f0702b75f924
https://github.com/snes9xgit/snes9x/commit/27fa8ecb972852e4b0100e5d3b266ec412a5ffad
https://github.com/snes9xgit/snes9x/commit/fc224c0df751e935ac5161e13b5f9be603953de7
https://github.com/snes9xgit/snes9x/commit/a66b88b276bd7f31a76a258d0d31b28cafef2d45

**Comment by hizzlekizzle** [comment]
oh wow, i didn't realize msu1 wasn't in upstream at all :O 

Thanks for the links!

**Comment by qwertymodo** [comment]
Well, it's in upstream, but 1.55 hasn't been released yet.

**Comment by hizzlekizzle** [comment]
oh gotcha. it's post-1.54/pre-1.55.

**Comment by hizzlekizzle** [comment]
Alright, @prataczak, I pulled in the apu, msu1 and stream files from upstream. See if that treats you any better. :)

**Comment by qwertymodo** [comment]
Since it looks like you pulled in the stream support, you should probably also pull in the rest of the .msu1 pack support as well.

https://github.com/snes9xgit/snes9x/commit/b22f5b5c442f42a8408ff65f08907c0281be9f9f
https://github.com/snes9xgit/snes9x/commit/f8da1524414209e4e090c3ec45aee59999c9aaf3
https://github.com/snes9xgit/snes9x/commit/ab2da6d9136ce9da559bcff47e47ef71363af11c
https://github.com/snes9xgit/snes9x/commit/9e804b4515520f7251866a07ade9c544d4ebb2d1
https://github.com/snes9xgit/snes9x/commit/03bdbe08fb3518048d081085e725d5797e28753c
https://github.com/snes9xgit/snes9x/commit/698bf1a4b2470a2aa0e98bd4fc282542b982d711
https://github.com/snes9xgit/snes9x/commit/158b509bbbbb979dc8a7789b60648fd9473f3de3
https://github.com/snes9xgit/snes9x/commit/7455501db8f6966b5d9e97d4c6ef783286f954e5
https://github.com/snes9xgit/snes9x/commit/cb3ace471e9fa73e80006da8364b5e800206b55b

If you want a good test case, take an unmodified copy of A Link to the Past and download http://www.mediafire.com/?qasv12obaaob49b into either the same directory as the ROM or into the patch directory (assuming RA has a soft-patch directory, not sure how you guys handle that), then simply rename the ROM file to match and then load the ROM.  If everything is working, you should get a soft-patched game with all of the MSU-1 goodness.

**Comment by prataczak** [comment]
I updated to the latest version, and tested with Street Fighter II Turbo MSU, and the skipping issue seems to be gone - however, when I would reset the emulator (to try to test again), the track that was playing before the reset would continue to play until a new track would take over.

So that's a little weird.  Definite improvement, but still not perfect. :)

**Comment by qwertymodo** [comment]
Looks like you're missing both instances of S9xResetMSU() in cpu.cpp

https://github.com/snes9xgit/snes9x/blob/master/cpu.cpp#L296
https://github.com/snes9xgit/snes9x/blob/master/cpu.cpp#L333

**Comment by hizzlekizzle** [comment]
Alright, I added in the reset lines to cpu.cpp and that seems to have taken care of it (once the PR is merged). I also took a stab at getting the compressed pack loading, but it seems that's going to be above my paygrade, so to speak. Typically, we leave the patching to the frontend rather than the core, so it may not be feasible at all.

**Comment by prataczak** [comment]
Thanks to you both. This is awesome stuff, and I can't wait to make RetroArch SNES9x core my default for this. :)

**Comment by qwertymodo** [comment]
Ah, I see now that all of the patching code has been removed from memmap.cpp, so yeah... that's not really mergeable in its current state.  I'd be interested to see how it's currently behaving without that... if anybody wants to test, grab that .msu1 file, open it as a .zip (rename it, or just open it with something like 7-zip that can heuristically determine the type), and grab the patch.bps file and hard-patch the LttP ROM (using beat or FloatingIPS), then rename the files to match (making sure to revert the file extension to .msu1), and see if the .pcm audio at least loads.  Being able to load MSU-1 packs without having to deal with 50 separate files is pretty convenient.

**Comment by prataczak** [comment]
I'm a little lost here, tbh... I run all my MSU games the same way I do in standalone SNES9x, I ditch the BML files that SNES9x doesn't use, then rename all of the other files the same thing (except the file extensions and the -1, -2, etc. for the .pcm files). Doesn't matter what. Then I put them in a folder named after whatever the name of the game is, and everything just seems to magically work. :)

I don't understand why Byuu ever made it as complicated as it was in the beginning but as it stands it's super easy with SNES9x. :)

**Comment by qwertymodo** [comment]
It's actually just as easy in higan, it wasn't byuu that made it complicated, it was MSU-1 hackers who wanted to release things in a way that worked everywhere.  higan doesn't actually need manifests, you can do it exactly like Snes9x, but instead of gamename-1.pcm, gamename-2.pcm, etc. you use track-1.pcm, track-2.pcm, and the gamename.msu file gets named msu1.rom.  Voila, manifest-free MSU-1 in higan.

My suggestion above is just one step easier, as all of the files are contained in a single .zip file and it all loads automatically without extracting, saving both time and disk space.

**Comment by prataczak** [comment]
Interesting... I must confess, I'm rather ignorant of how all of this stuff works. :)

I spoke out of place also. I have a lot of respect for Byuu, and higan/BSNES is certainly incredible stuff.

Give me some time, I'll see if I can make your suggestion work.

**Comment by qwertymodo** [comment]
Here's a quick visual comparison of the different configuration options: https://imgur.com/a/i7gmo  The last one is the relevant one here.  If they add soft-patching support, then the ROM can be an original unmodified ROM since the patch is actually contained inside the .msu1 pack along with the audio files, but since they don't support it yet, you'll have to extract the patch and apply it manually as I outlined above.  If you feel like it, give it a try in my Snes9x build and you can make that last one work without the extra patching step.

**Comment by prataczak** [comment]
I did as you instructed, and pulled the .bps file, and patched the zelda3 ROM with flips.exe.
I'm assuming I did this right, but here's a screen shot.
![2017-10-19 1](https://user-images.githubusercontent.com/12243933/31755444-7eac10b0-b46c-11e7-8fae-03770cef0e9e.png)

It seems to just be using the regular music.

**Comment by prataczak** [comment]
To verify, I tried loading the SFC file in SNES9x, and the alternate music played normally.

To make sure I understand this correctly, if this is implemented perfectly, I will have a folder with zipped ROMs, and in the same folder, all of the .msu1 files (named after the ROM counterparts), and everything will just work?

**Comment by qwertymodo** [comment]
Ok, found the likely culprit.

This: https://github.com/libretro/snes9x/blob/master/memmap.cpp#L2539  should be changed to this: https://github.com/snes9xgit/snes9x/blob/master/memmap.cpp#L2589 and then you can delete the MsuRomExists function from memmap.cpp.  You probably also need to include msu1.h.  I think in upstream I put that include into getset.h, which took care of it, but that might have been unnecessary, you might just include it in memmap.cpp directly.

**Comment by kps501** [comment]
snes9x upstream made lots of recent msu1 changes, and some bugfixes also. Not sure if this issue is resolved or not.

**Comment by ToniBC** [comment]
I am testing many games with MSU-1 in the latest versions of Snes9x and the sound is distorted in many effects (crackling). The same game in Higan does not fail, so it seems that the code of MSU-1 is with some error or not updated.

One of the games where it is best seen is Battletoads and Battlemaniacs, but all have errors in the sound, some in certain effects, but they all have something, like an excess of volume that distorts it.

I have also seen that MSU-1 games do not start with OpenGL but with Vulkan they work well.

**Comment by qwertymodo** [comment]
If you're hearing excessive volume, the first thing is to ensure that the audio files themselves are not actually too loud.  It's not as likely to be an issue these days, but it is still potentially possible.  I have no idea why the video driver would have any impact on the MSU-1.

**Comment by ToniBC** [comment]
As I mentioned, it is not a problem of the files, since the same ones in Higan do not present problems. MSU-1 gives support for music tracks and videos, but apparently when activated everything happens through MSU and as I mentioned, one of the most notorious is Battletoads, when the characters appear at the beginning, the distorted sound effect is heard .

That same game, I opened it with Higan with the corresponding manifest file and it does not fail. It is not a problem of the port of libretro, because in standalone the same happens and it is commented in several places, is that the implementation of the chip has some error in Snes9x.

OpenGL will try again, but being a different chip, outside of what is a real Snes, you may have a problem or is a matter of the frontend itself.

**Comment by stellarporter** [comment]
@ToniBC 
Upstream made some more improvements again in the last month or so. With better mixing added recently. Maybe give it another go again?

**Comment by ToniBC** [comment]
Yes, sorry, as they told me to put it as a separate failure, because I commented there. It seems that everything works well or the same as in Bsnes. It was a problem of excess volume. Now all sound good.

**Comment by hizzlekizzle** [comment]
ok, sounds like we can close this, then. Feel free to reopen if not.
